Bug 1814592 - Instrument the time needed for a successful/failure upload of a ping & upload shutdown in testing

badboy opened the pull request:
This PR does multiple things:

* Bring back #2365 -- that's the first 3 commits, cherry-picked unchanged (#2374 reverted them)
* Refactors the uploader shutdown to be used in test reset 
* Implement uploader shutdown in Swift
  * This is only used in test mode, because `glean::shutdown` is not called by the iOS bindings
  * Tries to be a bit more defensive for the newly introduced timing distributions done in the uploader.

Out of 20+ runs locally I had only 3 test failures ("No database found"). I don't fully understand why that still happens.
But hey, with only #2365 we had 7 tests fail basically every time. Improvement!

Especially the last part though bugs me. The iOS uploader still has some runs where it does some work after we told it to shut ~~up~~ down. And yet it happens in some tests.
Because Glean Swift _does not_ shut down in production use (it either runs or the embedding app is stopped/killed), this should not occur.
If it _does_ occur though this is strictly better by NOT crashing, but just logging an error now (but also: we only cover the known cases of metric types Glean collects there itself).

Time-wise: I'm still around on Monday, so we can have that reviewed and merged before.
I'd rather not have #2332 shipping without instrumentation, but it's also essentially desktop only for the time being in `main`, so it could.

I haven't (yet) tested this exact code on Desktop, but a prior version at least compiled in m-c just fine and ran tests.

badboy commented on glean-core/ios/Glean/Net/HttpPingUploader.swift:
```diff
@@ -4,6 +4,27 @@
 
 import Foundation
 
+/// `true` to allow the uploader to process pings.
+/// Note that this does not mean that an uploader is actually running.
+/// It will be invoked when a ping is submitted.
+///
+/// `false` to stop the uploader.
+var stateRunning: AtomicBoolean = AtomicBoolean(false)
+
+/// The uploader will enter and leave this group as it starts and finishes work.
+let uploadGroup = DispatchGroup()
+
+func shutdownUploader() {
+    stateRunning.value = false
+    // Wait for an uploader to finish it's current upload.
+    // If no uploader is running, this returns immediately.
+    uploadGroup.wait()
+}
+
+func startUploader() {
+    stateRunning.value = true
+}
```
oh this whole thing? more mutable global state!
I'm considering moving this into `HttpPingUploader` below and have it be a singleton like `Glean`

badboy commented:
> Base: **28.57**% // Head: **26.31**% // Decreases project coverage by **`-2.26%`** ⚠️

Ah yeah, new metrics, but no tests for those in Glean. Didn't I add one or did that get lost in my rebasing? Hmmm

chutten commented on glean-core/src/lib.rs:
```diff
@@ -514,37 +563,7 @@ pub fn shutdown() {
-            let _ = handle.join().ok();
-        }
-    }
+    uploader_shutdown();
```
Musing: I was reading through `uploader_shutdown` wondering if we could indeed call `with_glean` safely, and then get out here and read up-method to find that we do guard against init being fully complete, so that's helpful... but oof, this shutdown stuff is getting more and more complex and I'm not sure where the ideal place is to document or manage this complexity.

badboy replied:
ugh, yeah, it's ... a lot and spread in several places. This could definitely use some refactoring to streamline it again.

chutten commented on glean-core/src/lib.rs:
```diff
@@ -941,6 +960,13 @@ pub fn glean_test_destroy_glean(clear_stores: bool, data_path: Option<String>) {
 
         dispatcher::reset_dispatcher();
 
+        // Only useful if Glean was fully initialized.
```
Possible inconsistency: Is there a reason "Glean was fully initd" is determined differently here than at the top of shutdown()?

badboy replied:
Inconsistency indeed. `was_initialize_called` vs. "was called, did work and actually finish, including setting up a database"

chutten commented on glean-core/src/metrics/timing_distribution.rs:
```diff
@@ -216,9 +224,14 @@ impl TimingDistributionMetric {
+        // but in tests that storage might be gone already.
+        // Let's just ignore those.
+        // We do the same for counters.
+        // This should never happen in real app usage.
```
Suggestion: ...which means we can log it so we'll have a chance at being told?

chutten commented on glean-core/src/upload/mod.rs:
```diff
@@ -360,6 +365,18 @@ impl PingUploadManager {
+                    document_id,
+                    path
+                );
+                return;
```
Suggestion: This `return;` could be a big deal. We should have a health metric for it and a place on the monitoring dashboards for it.

chutten commented on glean-core/src/upload/mod.rs:
```diff
@@ -575,6 +592,13 @@ impl PingUploadManager {
 
+                {
+                    let mut in_flight = self.in_flight.write().unwrap();
+                    let success_id = self.upload_metrics.sending_success.start_sync();
+                    let failure_id = self.upload_metrics.sending_failure.start_sync();
```
Musing: I wonder if we should have safety notices for synchronous I/O. Like "Sync I/O: Acceptable because it's on the uploader thread which itself does waiting I/O on the network" or something. Might not be a problem until someone finds us on a perf profile.

chutten commented on glean-core/src/upload/mod.rs:
```diff
@@ -656,14 +680,27 @@ impl PingUploadManager {
+                        .sending_success
+                        .set_stop_and_accumulate(glean, success_id, stop_time);
+                    self.upload_metrics.sending_failure.cancel_sync(failure_id);
+                }
```
Suggestion: If `send_ids.is_none()` then that means 🤔  we've removed this doc_id before we got here, or we never inserted it in the first place... either way, that's an inconsistency. Perhaps we should instrument this?

badboy replied:
Instrument the instrumentation 🤔 I'm all for it.

chutten reviewed (approved):
Obviously also want Travis on here, but broadly I'm in favour of the approach and implementation.

badboy commented on glean-core/ios/Glean/Net/HttpPingUploader.swift:
```diff
@@ -159,19 +186,20 @@ public class HttpPingUploader {
                     Dispatchers.shared.launchAsync {
                         HttpPingUploader(configuration: self.config, testingMode: self.testingMode).process()
                     }
+                    uploadGroup.leave()
```
This only solves half the problem though.
In this `case` we're within the closure used by `self.upload`.
In the `.wait` and `.done` further below (lines 194ff) we're outside of it. 

So regardless we wil need one defer in there, but still the individual calls further below (because we don't want to double-fire it)

badboy commented:
I addressed all comments and filed bugs.
I also found one more thing: The metrics were named differently in `metrics.yaml` then they were in code. I fixed that ([bug 1815213 when?](https://bugzilla.mozilla.org/show_bug.cgi?id=1815213))
I also re-wrote the `upload_timing` test to use the metric test APIs instead of parsing the JSON (this should also satisfy codecov, because we had tests, it just didn't know about it)

badboy commented:
![image](https://user-images.githubusercontent.com/2129/218461211-8a85e0dc-d25c-4935-bed4-4363cbd4b0e1.png)

yey
